### PR TITLE
nixos/mailman: postorius & hyperkitty support

### PR DESCRIPTION
**Pull request by peti**
A simple Apache configuration for running Postorious and Hyperkitty looks as follows:

~~~
httpd = let webRoot = config.services.mailman.webRoot; in {
  enable = true;
  adminAddr = "postmaster@example.org";
  hostName = "lists.example.org";
  logPerVirtualHost = true;
  extraModules = [ { name = "wsgi"; path = "${pkgs.apacheHttpdPackages.mod_wsgi3}/modules/mod_wsgi.so"; } ];
  extraConfig = ''                                                                                                                                                                                                  
    WSGISocketPrefix /run/httpd/wsgi                                                                                                                                                                                
    WSGIDaemonProcess mailman threads=25 home=/var/lib/mailman-web python-path=${webRoot}:${
      lib.makeSearchPath pkgs.python3.sitePackages
        pkgs.python3Packages.mailman-web.requiredPythonModules
    }                                                                       
  '';
  virtualHosts = [
    { hostName = "lists.example.org";
      documentRoot = config.security.acme.certs."lists.example.org".webroot;
      extraConfig = ''                                                                                                                                                                                              
        <Directory "${config.security.acme.certs."lists.example.org".webroot}">                                                                                                                                     
          Options -Indexes                                                                                                                                                                                          
        </Directory>                                                                                                                                                                                                
        RedirectMatch permanent ^(?!/\.well-known/acme-challenge/).* https://lists.example.org$0                                                                                                                    
                                                                                                                                                                                                                    
      '';
    }
    { hostName = "lists.example.org";
      enableSSL = true;
      sslServerCert = "/var/lib/acme/lists.example.org/fullchain.pem";
      sslServerKey = "/var/lib/acme/lists.example.org/key.pem";
      servedDirs = [ { dir = "/var/lib/mailman-web/static"; urlPath = "/static"; } ];
      extraConfig = ''                                                                                                                                                                                              
        <Directory "${webRoot}">                                                                                                                                                                                    
          <Files wsgi.py>                                                                                                                                                                                           
            Require all granted                                                                                                                                                                                     
          </Files>                                                                                                                                                                                                  
          WSGIProcessGroup mailman                                                                                                                                                                                  
        </Directory>                                                                                                                                                                                                
        WSGIScriptAlias / ${webRoot}/wsgi.py                                                                                                                                                                        
      '';
    }
    { hostName = "localhost";
      servedDirs = [ { dir = "/var/lib/postorius/static"; urlPath = "/static"; } ];
      extraConfig = ''                                                                                                                                                                                              
        <Directory "${webRoot}">                                                                                                                                                                                    
          <Files wsgi.py>                                                                                                                                                                                           
            Require ip 127.0.0.1                                                                                                                                                                                    
            Require all granted                                                                                                                                                                                     
          </Files>                                                                                                                                                                                                  
          WSGIProcessGroup mailman                                                                                                                                                                                  
        </Directory>                                                                                                                                                                                                
        WSGIScriptAlias / ${webRoot}/wsgi.py                                                                                                                                                                        
      '';
    }
  ];
};
~~~


[1] https://gitlab.com/mailman/hyperkitty/tree/master/example_project

### CODE REVIEW

**Comment by peti** [comment]
I've been running this code on my server for a while now and consider it good enough(tm) to be useful to others.

File: pkgs/top-level/python-packages.nix [hunk 1]
```diff
@@ -715,6 +715,8 @@ in {
 
   mailman = disabledIf (!isPy3k) (callPackage ../servers/mail/mailman/core.nix { });
 
+  mailman-web = disabledIf (!isPy3k) (callPackage ../servers/mail/mailman/web.nix { });
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: This is not the right place. `python-packages.nix` should only contain libraries. This is configuration for the mailman module and should be situated there (it is) and be called from there (which it is not currently).

**Review by FRidh — CHANGES_REQUESTED** [review]

File: pkgs/top-level/python-packages.nix (hunk 1 above)
Comment (peti) [reply pkgs/top-level/python-packages.nix]: What do you mean by "there" exactly? Where should that definition be in your opinion?

Comment (FRidh) [reply pkgs/top-level/python-packages.nix]: in nixos/modules/services/mail/mailman.nix

Comment (peti) [reply pkgs/top-level/python-packages.nix]: I don't know ... I am not convinced that defining that module exclusively for NixOS is an improvement.

**Comment by FRidh** [comment]
It is not a Python module that is usable otherwise. The files are specific
to mailman, and they're also situated in a folder that they are not
importable.

**Comment by peti** [comment]
> It is not a Python module that is usable otherwise.

Users of Nixpkgs can use that module to set up the Mailman Web UI outside of NixOS.

**Comment by FRidh** [comment]
Great, then it needs to become a proper package. Modules go in
site-packages and toPythonModule needs to be used  so it identifies as a
Python module.

**Comment by peti** [comment]
> Great, then it needs to become a proper package.

I am sorry, @FRidh. I am aware of the fact that you dedicate a lot of time and effort into maintaining the Python infrastructure. I also realize that you perform a lot of updates, make sure everything works, and are actively maintaining that part of Nixpkgs here on Github. I totally respect that and I'd like to use the occasion to say *thank you* for all those efforts for all those years, which have benefited me greatly.

Having said that, the issue we are discussing here is a trivial minutiae that seems totally inconsequential. I don't understand where do you want to go with this. You seem to be pulling all kinds of rules and conventions out of thin air just for the sake of having something to criticize.

**Comment by FRidh** [comment]
@peti initially it did not make sense to me to have `mailman-web` in there because it appeared to be configuration only for a module. You then made the point that it actually can be reusable, and I agree now with that. We actually do this in other places as well (I recall jupyter notebook or lab), where the logic is not included in the module so it can be used elsewhere.

Right now, the issue is that the modules are not in a `site-package` folder. That means that, by default, they are not importable. To make them importable they should be in a `site-packages` folder, or an environment variable like `PYTHONPATH` needs to be used to make them importable. In the module, you made it work because `PYTHONPATH` is set using the `--pythonpath` attribute.

The Python packages is meant only for packages that provide importable packages and modules. That means the modules need to be in a `site-packages` folder and the derivation marked in a certain way, using `toPyhonModule` (which `buildPythonPackage`) already does. Functions like `python.buildEnv` and `python.withPackages` rely on this. This is the part that matters to me and why I have this discussion.

Now, Django configuration is typically not put in a default `site-packages` folder because on systems the dependencies are often put separate from the web application configuration, however, there is no harm in doing so.

**Comment by peti** [comment]
OK. That makes sense. I'll make the necessary changes to `mailman-web`.

File: pkgs/servers/mail/mailman/web.nix [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  allowedHostsString = stdenv.lib.concatMapStringsSep ", " (x: "\""+x+"\"") allowedHosts;
+
+in
+
+stdenv.mkDerivation {
+  name = "mailman-web-0";
+
+  unpackPhase = ":";
+
+  installPhase = ''
```
Comment (FRidh) [inline pkgs/servers/mail/mailman/web.nix]: these could be moved to `$out/${python.sitePackages}`

File: pkgs/top-level/python-packages.nix (hunk 1 above)
Comment (peti) [reply pkgs/top-level/python-packages.nix]: I turned mailman-web into a proper python package in 86f8895abbf.

File: pkgs/servers/mail/mailman/web.nix (hunk 2 above)
Comment (peti) [reply pkgs/servers/mail/mailman/web.nix]: 86f8895abbfdb3ce118e0ca971f56c7f72871e29

**Review by FRidh — APPROVED** [review]

**Pull request merged**